Add loongarch support

yetist opened the pull request:
这是移植了所有功能的补丁

xen0n commented:
See https://github.com/loongarch64/systemd/pull/3#issuecomment-909190675

yetist committed:
seccomp: add LoongArch 64bit support

yetist commented:
这是最新全功能补丁，接下来会拆分出来一个一个向上游提交。

xen0n commented on src/basic/architecture.h:
```diff
@@ -198,10 +199,30 @@ int uname_architecture(void);
+#    else
+#      error "Unrecognized loongarch architecture variant"
+#    endif
+#  else
```
明确判断 `__loongarch32` 然后 `#else` 报错吧，别哪天有人心血来潮照着 RV128 抄个 LA128 就傻逼了

然后 32 放 64 前面会好一些

xen0n commented:
冲突了 解一下

yetist commented:
都验证过了，这个 PR 整体没问题，但是由于 `gnu-efi`（影响boot) 和 `libseccomp` 还没有合并到上游，所以应该把这3次提交拆分了提交上游。

xen0n commented:
只剩俩补丁没上游了，很感动

yetist commented:
> 只剩俩补丁没上游了，很感动

话说你那个 nspr 该动一下了， :laughing:

xen0n commented:
要不要把 `__loongarch64` 都改 `__loongarch_lp64` 或者 `__loongarch__ && __loongarch_grlen == 64`

另外要不要跟进 `loongarchNN-linux-gnuf64` &rarr; `loongarchNN-linux-gnu` 也是个问题（如果要改的话尽量带上讨论链接，算是尊重历史真相了，也是给 systemd 的人一点上下文）

yetist commented:
> 要不要把 `__loongarch64` 都改 `__loongarch_lp64` 或者 `__loongarch__ && __loongarch_grlen == 64`
> 
> 另外要不要跟进 `loongarchNN-linux-gnuf64` → `loongarchNN-linux-gnu` 也是个问题（如果要改的话尽量带上讨论链接，算是尊重历史真相了，也是给 systemd 的人一点上下文）

有道理，`__loongarch64`  是用于兼容早期移植代码的 C/C++ 预处理器内建宏。

yetist commented:
这个PR应该还得留在这里一段时间，直到gnu-efi和libseccomp发了新版本之后，才能发起合并。

xen0n commented on src/boot/efi/meson.build:
```diff
@@ -284,9 +285,9 @@ else
+if efi_arch[1] in ['arm', 'riscv64', 'loongarch64'] or (efi_arch[1] == 'aarch64' and (not objcopy_2_38.found() or coff_header_in_crt0))
         efi_ldflags += ['-shared']
-        # ARM32 and 64bit RISC-V don't have an EFI capable objcopy.
+        # ARM32 and 64bit LoongArch and RISC-V don't have an EFI capable objcopy.
```
当时你做这变更时候 LA UEFI 确实没 `objcopy` 可用（binutils 没支持 LA 架构的 COFF 格式），[但是后来有了](https://github.com/bminor/binutils-gdb/commit/31f6009538b75c22a4b3f99216846973283f0c6d)，binutils 2.40 应该已经可用

跟进下？

yetist commented:
就差seccomp了，boot上游自己加上了，不过新依赖pyelftools（补丁已经合并）。

ddstreet committed:
openssl: add log_openssl_errors()

Add a macro to log all errors in the openssl 'thread error queue'.

This consolidates all the openssl-generated errors to return -EIO and log at
debug level.

Also add a 'invalid' test in test-openssl, to allow manual verification that
the openssl error(s) are logged.

rpigott committed:
zsh: use sys_really_all_units for non-template names

The systemctl invocations used for these completions match the ones used
for the _sys_really_all_units parameter, so we should really just use
the cached parameter rather than recomputing the result.

yuwata committed:
meson: use install_emptydir() and drop meson-make-symlink.sh

The script is mostly equivalent to 'mkdir -p' and 'ln -sfr'.
Let's replace it with install_emptydir() builtin function and
inline meson call.

yuwata committed:
udev: set ID_NAME and ID_SERIAL to MMC/memstick devices again

Fixes a bug introduced by 998db5871fea331ec00b26a3a3f5271df040a905.

Fixes #28671.

rpigott committed:
zsh: reintroduce pattern argument to uncached verbs

The systemctl completion previously made use of PREFIX as a pattern
argument to list-unit-files and list-units. This had the problem of
erroneously filtering the results that were stored in the cache, and
erroneously filtering results that might have been requested according
to the users configuration (e.g. _correct completer, certain
matcher-lists or tag-orders, etc.).

Unfortunately, the runtime of list-unit-files increases when no pattern
argument is provided, and systemctl show, used to filter those units,
can become unacceptably slow when provided with too many units to
describe.

Let's re-introduce the pattern argument to list-unit-files and
list-units where necessary in order to alleviate these bottlenecks
without poisining the cache. A 'use-pattern' style is introduced that
may be used to disable this behavior if it is undesired. We can still
expect that certain completions, like `systemctl start <TAB>` will be
slow, like before. To fix this we will need systemd to learn a more
efficient way of filtering the units than parsing systemctl show.

poettering committed:
libfido2: pick up debug logging from libfido2, and funnel it through our log subsystem

Fixes: #27984

Vogtinator committed:
units/initrd-parse-etc.service: Conflict with emergency.target

If emergency.target is started while initrd-parse-etc.service/start is queued,
the initrd-parse-etc job did not get canceled. In parallel to the emergency
units, it eventually runs the service, which starts initrd-cleanup.service,
which in turn isolates initrd-switch-root.target. This stops the emergency
units and effectively starts the initrd boot process again, which likely
fails again like the initial attempt. The system is thus stuck in an endless
loop, never really reaching emergency.target.

With this conflict added, starting emergency.target automatically cancels
initrd-parse-etc.service/start, avoiding the loop.

bluca committed:
Merge pull request #28710 from rpigott/zsh-non-template-cache

zsh completion speedups

copperii committed:
po: Translated using Weblate (Finnish)

Currently translated at 86.3% (196 of 227 strings)

Co-authored-by: Jan Kuparinen <copper_fin@hotmail.com>
Translate-URL: https://translate.fedoraproject.org/projects/systemd/master/fi/
Translation: systemd/main

yuwata committed:
tree-wise: drop unnecessary use of proc_cmdline_key_streq()

If the key does not contain '-' or '_', then it is not necessary to use
proc_cmdline_key_streq(), and streq() is sufficient.

This also adds missing assertions about 'key' argument.

yuwata committed:
resolve: ignore nameserver= and domain= kernel command line options without value

Otherwise, manager_parse_dns_server_string_and_warn() or
manager_parse_search_domains_and_warn() will trigger assertion.

mordner committed:
man: fix typo in journalctl

bluca committed:
Merge pull request #28727 from yuwata/proc-cmdline-cleanups

tree-wide: fixlet and optimization for parsing kernel command line

yetist committed:
LoongArch: Replace __loongarch64 with __loongarch_lp64

`__loongarch64` is no longer used for historical reasons and should be replaced by `__loongarch_lp64` in new code.

https://loongson.github.io/LoongArch-Documentation/LoongArch-toolchain-conventions-EN.html#_cc_preprocessor_built_in_macro_definitions
